_drafts/floating-point-101.md

```diff
@@ -0,0 +1,444 @@
+image: /assets/floating-point-101/banner.png
+stylesheet: /assets/floating-point-101/number-line.css
+---
+
+Software engineering keeps getting more abstract, but one thing is unchanging: the importance of floating-point
+arithmetic. _Every_ computer programmer is bound to work with numbers (they call them _computers_ for a reason), so it's
+genuinely useful to understand the way machines do math, no matter if your code is for a to-do app, a stock exchange, or
+a fridge. How are numbers stored exactly? What's the significance of special values? And why is 0.1 + 0.2 not equal to
+0.3? Let's explore this all!
+
+## The Standard
+
+Let's start with one key assumption: in all the world, on every continent, there's one and one only way of doing
+floating-point arithmetic.
+
+This makes things a lot easier, _and_ it's true in practice. That standard's name?
+
+~~Albert Ein~~ **IEEE 754**.
+
+How's this so simple? I mean, we all know how attempts at standardization _really_ turn out.
+
+<figure>
+<a href="https://xkcd.com/927/"><img src="https://imgs.xkcd.com/comics/standards_2x.png" title="Fortunately, the charging one has been solved now that we've all standardized on mini-USB. Or is it micro-USB? Shit." alt="Standards" height="283" width="500" loading="lazy"></a>
+<figcaption><i>fig. 1</i> — There's always a relevant XKCD.</figcaption>
+</figure>
+
+You'd be correct to think more than one format must have been invented. _Plenty_ were – in the early days of computing,
+practically every system with floating-point capabilities had its own. Later on, brand-specific formats emerged: IBM
+went for hexadecimal floating-point in their mainframes, Microsoft created Microsoft Binary Format for their BASIC
+products, DEC cooked up yet something else for the VAX architecture.
+
+This changed when Intel decided in the late 70s to design the floating-point chip to rule them all – which required a
+format to rule them all too, the best possible. The decision culminated in the Intel 8087 coprocessor of 1980, but even
+before that, other companies in the space caught wind of this work and set up
+[a common effort](https://www.researchgate.net/publication/2954891_IEEE_754_An_Interview_with_William_Kahan) at the
+Institute of Electrical and Electronics Engineers (IEEE) to standardize floating-point arithmetic – the IEEE 754 working
+group. Two competing drafts prevailed: the Intel 8087 spec vs. the DEC VAX one. After some more arguments and error
+analysis, in 1981 Intel's draft won out, rapidly got adopted by everyone,
+and [the rest is history](https://www.intel.com/content/dam/www/public/us/en/documents/case-studies/floating-point-case-study.pdf)
+(though it took the committee another four years of bickering to publish that draft, of course).
+
+> For a detailed look at historic floating-point formats, see
+> [this great article by John Savard](http://www.quadibloc.com/comp/cp0201.htm).
+
+## The Specs
```

Comment:
I'd rather begin here, with something like:

"To do anything with floating point numbers, you first need a way to represent them"

And contrast with how we represent integers, and how/why floating points of arbitrary precision aren't that straight forward, for whatever reason.

And somewhere in the middle, I'd put a para for what it would take for infinite precision, and why it isn't possible in practice, to drive the point home.

Reply:
That's a great suggestion, this did feel clunky to me, but I didn't have a great idea of starting in a more useful way.